File: data_structures/linked_list/linked_list.rb
class LinkedList
  attr_reader :length

  def initialize
    @head = Node.new
    @length = 0
  end

  def find(val)
    node = @head.next
    until node == nil || node.val == val
      node = node.next
    end
    node
  end

  def append(val)
    node = @head
    until node.next.nil?
      node = node.next
    end
    @length += 1
    node.next = Node.new(val)
  end

  def delete(val)
    node = @head.next
    until node.next.nil?
      if node.next.val == val
        node.next = node.next.next
      end
      node = node.next
    end
  end

  def to_s
    s = []
    node = @head.next
    until node.nil?
      s << node.val.to_s
      node = node.next
    end
    s = s.join(", ")
    "[" + s + "]"
  end

private

  class Node
    attr_accessor :val, :next

    def initialize(val = nil)
      @val = val
      @next = nil
    end
  end

end


x = LinkedList.new
x.append(1)
x.append(2)
x.append(3)
x.append(4)
x.delete(3)
puts x

File: data_structures/lru_cache/lru/linked_list.js
class LinkedList{

  constructor(){
    this.head = null;
    this.tail = null;
    this.length = 0;
    this.find = this.find.bind(this);
  }

  mostRecent(){
    this.tail.val;
  }

  oldest(){
    this.head.val;
  }

  append(key, val){
    let newNode = new Node(key, val);
    if(this.length === 0){
      this.head = newNode;
      this.tail = this.head;
    }
    else if(this.length === 1){
      let tail = newNode;
      tail.prev = this.head;
      this.head.next = tail;
      this.tail = tail;
    }
    else{
      let prevNode = this.tail;
      this.tail = newNode;
      prevNode.next = this.tail;
      this.tail.prev = prevNode;
    }
    this.length++;
  }

  find(key){
    let node = this.head;
    for (var i = 0; i < this.length; i++) {
      if(node.key === key) return node;
      node = node.next;
    }
    return null;
  }

  delete(key){
    let delNode = this.find(key);
    if (delNode === null) return null;

    if(this.length === 1){
      this.head = null;
      this.tail = null;
    }
    else if(this.head === delNode){
      this.head = this.head.next;
      this.head.prev = null;
    }
    else if(this.tail === delNode){
      this.tail = this.tail.prev;
      this.tail.next = null;
    }
    else{
      delNode.prev.next = delNode.next;
      delNode.next.prev = delNode.prev;
    }
    this.length--;
  }

  toString(){
    let s = "";
    let node = this.head;
    for (var i = 0; i < this.length; i++) {
      s += `${i} - key: ${node.key} - val: ${node.val}`;
      node = node.next;
    }
    return node;
  }


}

class Node{
  constructor(key, val){
    this.key = key;
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

let list = new LinkedList;
list.append(1, 5);
list.append(2, 3);
list.append(3, 4);
console.log(list);
console.log(list.length);
list.delete(2);
console.log(list);
console.log(list.length);
